src/p2p/swarm.rs

```diff
@@ -38,11 +38,18 @@ type NetworkBehaviourAction = swarm::NetworkBehaviourAction<<<<SwarmApi as Netwo
 #[derive(Debug, Default)]
 pub struct SwarmApi {
     events: VecDeque<NetworkBehaviourAction>,
+
+    // FIXME: anything related to this is probably wrong, and doesn't behave as one would expect
+    // from the method names
     peers: HashSet<PeerId>,
     connect_registry: SubscriptionRegistry<(), String>,
     connections: HashMap<MultiaddrWithoutPeerId, PeerId>,
     roundtrip_times: HashMap<PeerId, Duration>,
     connected_peers: HashMap<PeerId, Vec<MultiaddrWithoutPeerId>>,
+
+    /// The connections which have been requested, but the swarm/network is yet to ask for
+    /// addresses; currently filled in the order of adding, with the default size of one.
+    pending_addresses: HashMap<PeerId, Vec<Multiaddr>>,
```

Comment:
I was wondering if a `HashSet` wouldn't be better than having to do linear lookups, but I guess a `Vec` should be faster for only a few entries, and I wouldn't expect more :+1:

Reply:
Also the use in `NetworkBehaviour::addresses_of_peer` which is just a `pending_addresses.remove(peer_id).unwrap_or_default()`.

This is never expected to accumulate a lot of addresses. Perhaps there's a chance somehow to accumulate dead entries here, like getting `libp2p_kad` dialing to the same peer at the sibling poll. Even in that case, the hashmap entry would get removed. Perhaps not in all cases. Overall it'd be easier to get access to the networking core of the Swarm for all of these, but at the same time I kind of agree hiding it is a good idea. Perhaps having access to `HashSet<PeerId>` via PollParameters would work. Need to still keep thinking this :)

Reply:
I'll add cleanup to addr_reach_failure and disconnect at least.

Reply:
Again, it seems that this is not so simple. I guess I'll have to follow up with another PR to see if this could just be "connect to the peer with this given multiaddr, or any other way". That should at least simplify a lot of the wrestling, and allow to use the non-provided methods of NetworkBehaviour.

Noted an additional "deadlock" with a peer racing to connect to us before our dialing attempt has completed. This should be solvable by redefining the `connect` as "end up being in connection to a peer."

Reply:
This is quite outdated. Might be best to look again with fresh eyes :)